### DISCUSSION TITLE
Manipulating a non-existent field

### DISCUSSION THREAD

**Discussion by elavy-harris** [nextgenhealthcare/connect#6266] [Q&A] [answered]
I've written the following function to ensure that fields don't get created in the XML out of order:

```
function padField(field, padLength) {
	const fieldName = field.name();
	for (component = 1; component <= padLength; component++) {
		if (!field[fieldName + '.' + component].toString()) {
			field[fieldName + '.' + component] = '';
		}
	}
}
```

I call it with something like `padField(msg['PV1']['PV1.7'], 9)`.

This works fine most of the time, but if there is no PV1.7 (e.g. PV1 only has the first 6 fields), it will fail on the `name()` call with the error `TypeError: name is not a function, it is xml.`

I can avoid the error by inserting `if (!field.toString()) return;`, but I really want to create all the components.

For the record, if I pass in a newly-created segment instead of a non-existent field, `toString()` is false, but `name()` works, i.e.
```
	let pv1 = createSegmentAfter('PV1',  msg['PID']);
	padField(pv1, 45);
```

Is there any way for this function to know how to do its job?

**Comment by rjkroll** [comment]
You could try something like this, which includes a check at the beginning to see if the field exists. If it doesn't, it creates the field and initializes the specified number of components. This ensures that the function can handle cases where the field is missing.
```
function padField(segment, fieldName, padLength) {
    if (!segment || typeof fieldName !== 'string' || !Number.isInteger(padLength) || padLength < 1) {
        throw new Error('Invalid input to padField function');
    }

    // Extract the field number from the fieldName (e.g., '7' from 'PV1.7')
    const fieldNumber = fieldName.split('.')[1];

    // Check if the field exists, if not, create it
    if (!segment[fieldName] || segment[fieldName].toString() === '') {
        segment[fieldName] = ''; // Initialize the field
    }

    // Pad the field with empty components
    for (let component = 1; component <= padLength; component++) {
        const componentKey = fieldName + '.' + component;
        if (segment[componentKey] == null || segment[componentKey] === '') {
            segment[componentKey] = '';
        }
    }

    return segment[fieldName];
}
```
usage: padField(msg['PV1'], 'PV1.7', 9);    <--- or whatever segment

- Takes three parameters: the segment object, the field name, and the desired padding length.
- Works with any segment and field
- Creates the field if it doesn't exist.
- Pads the field with empty components up to the specified length.

**Reply by elavy-harris** [reply]
Thanks, @rjkroll .

There's a lot to your code that goes beyond the immediate question. Passing in `fieldName`, as opposed to computing it, seems to be the crux of the approach.

I can see how this would address my issue. I was hoping there was a way to derive `fieldName` reliably, without having to specify it explicitly.

**Reply by tonygermano** [reply]
> I can see how this would address my issue. I was hoping there was a way to derive `fieldName` reliably, without having to specify it explicitly.

When you pass it a field which does not exist, you are passing in an empty XMLList, and as far as I can tell there is no way to retrieve the original call path that resulted in the empty XMLList. Not only can you not call the `.name()` method, but `field.length()` is zero (coincidentally, that's what you should have been using instead of `!field.toString()` for your early return) and `field.parent()` is undefined. When you pass it an existing field, then `field.parent()` should return the segment.

**Reply by tonygermano** [reply]
I'm guessing AI wrote the function at the top of this thread because it had a ton of mistakes that wouldn't have allowed it to work. This is a mostly working version of it. It also shows an alternative for checking if something exists in xml. Instead of `segment[fieldName].length() > 0` you can also check `fieldName in segment`. Just be aware that the `in` operator has really low precedence, which is why I needed to wrap it in parenthesis to get the "not" value.
```javascript
function padField(segment, fieldName, padLength) {
    if (!segment || typeof fieldName !== 'string' || !Number.isInteger(padLength) || padLength < 1) {
        throw new Error('Invalid input to padField function');
    }

    // Check if the field exists, if not, create it
    if (!(fieldName in segment)) {
        segment[fieldName] = <{fieldName}/>; // Initialize the field (but we aren't checking for correct order here!)
    }

    // Pad the field with empty components
    for (var component = 1; component <= padLength; component++) {
        var componentKey = fieldName + '.' + component;
        if (!(componentKey in segment[fieldName])) {
            segment[fieldName][componentKey] = <{componentKey}/>;
        }
    }

    return segment[fieldName];
}
```

**Reply by elavy-harris** [reply]
Thanks, @tonygermano .

The function I posted originally was fine for me, until I ran into the situation where the field didn't exist yet. Passing in `fieldName`, instead of deriving it, seems to be an adequate (although not elegant) solution. I'll look into using `in` or `length() === 0` in place of `toString()`.

> field.parent() is undefined

I found your statement to be true, but somehow the object is aware of its parentage anyway. I did the following
```
padTest(msg['PID']['PID.99'], 3, 'PID.99');
logger.info('pid.99=' + msg['PID']['PID.99'].toString());

function padTest(field, padLength, fieldName) {
	logger.info('toString=' + field.toString() + ', length=' + field.length() + ', parent=' + field.parent());
	for (component = 1; component <= padLength; component++) {
		if (!field[fieldName + '.' + component].toString()) {
			field[fieldName + '.' + component] = '';
		}
	}
	logger.info('toString=' + field.toString() + ', length=' + field.length() + ', parent=' + field.parent());
}
```

This code populates PID.99 successfully, without any initialization, and the `logger.info` at the bottom of the function displays the contents of the PID segment.

While there's redundancy here (saying 'PID.99' twice), I think I prefer it in this case over passing in the parent and name. In addition to using the method for creating components in a field, I also use it for creating fields in a segment. That segment COULD be one of a repetition, so the parent is `msg`, but I might need to be referencing `msg['PV1'][2]`.

**Reply by tonygermano** [reply, marked as answer]
Three main points -
1. I think the simplest solution is to use the code template I mentioned [in my other answer](https://github.com/nextgenhealthcare/connect/discussions/6266#discussioncomment-10155320).
2. You mentioned padding fields in a segment as well as components in a field. I think if you don't use option 1, then you should do this by default so that,
   - fields are created in the correct order, and
   -  you don't need to worry about a field not existing when go to add components
3. You made an interesting discovery, but I don't know if that's how e4x is supposed to work or it's just a quirk of the current Rhino implementation. **I personally would not rely on it**, but it turns out you can get the fieldName without passing it in separately. I think something like this is closer to what you originally were trying to do.
```javascript
function padTest(parent, padLength) {
    // create parent node if it does not exist in the current document
    if (parent.length() === 0) {
        parent.@temp = ''
        delete parent.@temp
    }

    const parentName = parent.localName()

    for (childId = 1; childId <= padLength; childId++) {
        var childName = parentName + '.' + childId
        if (!(childName in parent)) {
            parent[childName] = <{childName}/>;
        }
    }
}
```

However, this **still** has a problem that would be solved by using option 1.  Presumably this could occur because you check if a child already exists before creating it. For example,
- Your newly created field, PID-99, already has components 1 and 3 defined.
- You call `padTest(msg['PID']['PID.99'], 4)`
- The field will create components 2 and 4 **after** the third component.

Finally, a minor point, I've been using `parent[childName] = <{childName}/>` instead of `parent[childName] = ''`. Both of them will replace existing child elements of the parent if they already exist or create a new child at the end of the list if not. The former will assign a truly empty element. The latter will assign a new element with a child text node containing an empty string. To illustrate,
```javascript
js> var x = <xml/>
js> x.a = <a/>
js> x.b = ''
js> x.a['a.1'] = 'a.1'
a.1
js> x.b['b.1'] = 'b.1'
b.1
js> x.toString()
<xml><a><a.1>a.1</a.1></a><b><b.1>b.1</b.1></b></xml>
js> x.a.children().length()
1
js> x.b.children().length() // still contains the empty string text node
2
```

**Reply by elavy-harris** [reply]
@tonygermano For the question I was asking, I'm going to mark your prior comment as the answer, specifically for this chunk of code:
```
if (parent.length() === 0) {
  parent.@temp = ''
  delete parent.@temp
}
```

I had to modify it to use `parent.temp` instead of `parent.@temp`, because our linter doesn't like the "@". Since we've already determined that `parent` is empty, I don't think it matters what temp field we create and destroy.

I'll consider adopting your code template. For the record, I've got other use cases besides `padField()` that can also benefit from the "temp" logic, e.g. a function that processes provider fields so might be passed `msg['PV1']['PV1.7'][i]` or `msg['PV1']['PV1.8'][i]` or ....

Some questions, though:
1. What's the difference between `name()` and `localName()`? For my HL7 content, they both seem to return the same value.
2. Is there a significant difference between `field.length()===0` and `!field.toString()`?
3. How did you get a command-line environment that would correspond with Mirth's Rhino?

Other comments:
1. I don't know how it works, but the E4X behavior I described seems essential. Really, there's no difference between `msg['PID']['PID.99']['PID.99.1']='x'` and `let a=msg['PID']['PID.99'];a['PID.99.1']='x'`. We know the former works as desired. Based on prior discussion, if we were to evaluate `a` in the middle of the latter, we'd find no apparent connection to `msg['PID']`.

**Reply by elavy-harris** [reply]
My linter also doesn't like `<{childName}/>`. Is `new XML('<' + childName + '/>')` equivalent?

**Reply by pacmano1** [reply]
Don't use your linter?  (and what linter are you using?)

**Reply by elavy-harris** [reply]
> Don't use your linter? (and what linter are you using?)

JAVASCRIPT_ES

**Reply by pacmano1** [reply]
I can't think of a linter that supports e4x properly.

**Reply by tonygermano** [reply]
> My linter also doesn't like `<{childName}/>`. Is `new XML('<' + childName + '/>')` equivalent?

Yes, that is equivalent.

> msg['PV1']['PV1.7'][i]

This will not work, because as soon as you try to access an XMLList by index, it will evaluate to `undefined` instead of another empty XMLList when the index does not exist.


> What's the difference between `name()` and `localName()`? For my HL7 content, they both seem to return the same value.

`.name()` returns a QName object that has a `uri` string component for the namespace and a `localName` string component for the local name.  When the namespace is an empty string (which is the default for mirth hl7v2) then the toString() value of the QName object is the same as the localName. When the namespace is not empty, then `.name().toString()` returns "namespaceUri::localName".

`.localName()` just returns the localName string directly instead of an object and regardless of namespace.


> Is there a significant difference between `field.length()===0` and `!field.toString()`

`field.length() === 0` tells you definitively that `field` is an empty XMLList.

`!field.toString()` tells you that `field` may be an empty XMLList, or it already exists, but has a single element without any children except possibly empty string text nodes.

If you use `!field.toString()` and run `padTest` twice. All of the elements you created in the first run will be replaced in the second run due to having an empty toString() value.

**Reply by tonygermano** [reply]
I realize I left a couple things unaddressed, but that's all I have time for right now, and I figured this was better than nothing.

**Reply by elavy-harris** [reply]
> > msg['PV1']['PV1.7'][i]
> 
> This will not work, because as soon as you try to access an XMLList by index, it will evaluate to `undefined` instead of another empty XMLList when the index does not exist.

My use case here limits to pre-existing instances of PV1.7, e.g.
```
		for (let index = 0 ; index < msg['PV1']['PV1.7'].length() ; index++) 
		{
			processProvider(msg['PV1']['PV1.7'][index]);	
		}
```

For those, I know that the field exists and has a `name()`. It's when I do something like `processProvider(rol['ROL.4'])` that I don't.

**Comment by tonygermano** [comment]
This doesn't directly answer your question, but if you plan on adding fields out of order, I'd take a look at this code template. You can add fields or components in any order you want and have it clean up for you at the end of the transformer.

https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Fix%20HL7%20Node%20Order
